hudi-common/src/main/java/org/apache/hudi/common/table/timeline/TimelineDiffHelper.java

```diff
@@ -57,35 +57,38 @@ public static TimelineDiffResult getNewInstantsForIncrementalSync(HoodieTimeline
+      List<Pair<HoodieInstant, HoodieInstant>> viewChangingInstants = getPendingActionTransitions(oldT, newT);
+      List<HoodieInstant> lostPendingActions = viewChangingInstants.stream()
           .filter(instantPair -> instantPair.getValue() == null).map(Pair::getKey).collect(Collectors.toList());
-      if (!lostPendingCompactions.isEmpty()) {
-        // If a compaction is unscheduled, fall back to complete refresh of fs view since some log files could have been
+      if (!lostPendingActions.isEmpty()) {
+        // If a compaction/clustering is unscheduled, fall back to complete refresh of fs view since some log files could have been
         // moved. Its unsafe to incrementally sync in that case.
-        LOG.warn("Some pending compactions are no longer in new timeline (unscheduled ?). They are :"
-            + lostPendingCompactions);
+        LOG.warn("Some pending view changing instants are no longer in new timeline (unscheduled ?). They are :"
+            + lostPendingActions);
         return TimelineDiffResult.UNSAFE_SYNC_RESULT;
       }
-      List<HoodieInstant> finishedCompactionInstants = compactionInstants.stream()
-          .filter(instantPair -> instantPair.getValue().getAction().equals(HoodieTimeline.COMMIT_ACTION)
-              && instantPair.getValue().isCompleted())
+      List<HoodieInstant> finishedViewChangingInstants = viewChangingInstants.stream()
+          .filter(instantPair -> instantPair.getValue().isCompleted()
+              && (HoodieTimeline.COMMIT_ACTION.equals(instantPair.getValue().getAction())
+              || HoodieTimeline.REPLACE_COMMIT_ACTION.equals(instantPair.getValue().getAction())))
           .map(Pair::getKey).collect(Collectors.toList());
 
       newT.getInstants().filter(instant -> !oldTimelineInstants.contains(instant)).forEach(newInstants::add);
-      return new TimelineDiffResult(newInstants, finishedCompactionInstants, true);
+      return new TimelineDiffResult(newInstants, finishedViewChangingInstants, true);
     } else {
       // One or more timelines is empty
       LOG.warn("One or more timelines is empty");
       return TimelineDiffResult.UNSAFE_SYNC_RESULT;
     }
   }
 
-  private static List<Pair<HoodieInstant, HoodieInstant>> getPendingCompactionTransitions(HoodieTimeline oldTimeline,
-      HoodieTimeline newTimeline) {
+  private static List<Pair<HoodieInstant, HoodieInstant>> getPendingActionTransitions(HoodieTimeline oldTimeline,
+                                                                                      HoodieTimeline newTimeline) {
     Set<HoodieInstant> newTimelineInstants = newTimeline.getInstants().collect(Collectors.toSet());
+    
+    List<Pair<HoodieInstant, HoodieInstant>> allTransitions = new ArrayList<>();
 
-    return oldTimeline.filterPendingCompactionTimeline().getInstants().map(instant -> {
+    allTransitions.addAll(oldTimeline.filterPendingCompactionTimeline().getInstants().map(instant -> {
```

Comment:
Instead of oldTimeline.filterPendingCompactionTimeline().getInstants(),
is there scope to use 
oldTimeline.viewChangingInstants() and consolidate both this and below statement  where we handle replace commits ?